packages/glimmer-runtime/lib/compiled/opcodes/content.ts

```diff
@@ -89,11 +106,121 @@ export abstract class AppendOpcode<T extends Insertion> extends Opcode {
   }
 }
 
+export abstract class GuardedAppendOpcode<T extends Insertion> extends AppendOpcode<T> {
+  protected abstract AppendOpcode: typeof OptimizedCautiousAppendOpcode | typeof OptimizedTrustingAppendOpcode;
+  private expression: CompiledExpression<any>;
+  private deopted: OpSeq = null;
+
+  constructor(expression: CompiledExpression<any>) {
+    super();
+    this.expression = expression;
+  }
+
+  evaluate(vm: VM) {
+    if (this.deopted) {
+      vm.pushEvalFrame(this.deopted);
+    } else {
+      vm.evaluateOperand(this.expression);
+
+      let value = vm.frame.getOperand().value();
+
+      if(isComponentDefinition(value)) {
+        vm.pushEvalFrame(this.deopt(vm.env));
+      } else {
+        super.evaluate(vm);
+      }
+    }
+  }
+
+  public deopt(env: Environment): OpSeq { // Public because it's used in the lazy deopt
+    // At compile time, we determined that this append callsite might refer
+    // to a local variable/property lookup that resolves to a component
+    // definition at runtime.
+    //
+    // We could have eagerly compiled this callsite into something like this:
+    //
+    //   {{#if (is-component-definition foo)}}
+    //     {{component foo}}
+    //   {{else}}
+    //     {{foo}}
+    //   {{/if}}
+    //
+    // However, in practice, there might be a large amout of these callsites
+    // and most of them would resolve to a simple value lookup. Therefore, we
+    // tried to be optimistic and assumed that the callsite will resolve to
+    // appending a simple value.
+    //
+    // However, we have reached here because at runtime, the guard conditional
+    // have detected that this callsite is indeed referring to a component
+    // definition object. Since this is likely going to be true for other
+    // instances of the same callsite, it is now appropiate to deopt into the
+    // expanded version that handles both cases. The compilation would look
+    // like this:
+    //
+    //               Enter(BEGIN, END)
+    //   BEGIN:      Noop
+    //               Test(is-component-definition)
+    //               JumpUnless(VALUE)
+    //   COMPONENT:  Noop
+    //               PutDynamicComponentDefinitionOpcode
+    //               OpenComponent
+    //               CloseComponent
+    //               Jump(END)
+    //   VALUE:      Noop
+    //               OptimizedAppend
+    //   END:        Noop
+    //               Exit
+    //
+    // Keep in mind that even if we *don't* reach here at initial render time,
+    // it is still possible (although quite rare) that the simple value we
+    // encounter during initial render could later change into a component
+    // definition object at update time. That is handled by the "lazy deopt"
+    // code on the update side (scroll down for the next big block of comment).
+
+    let buffer = new CompileIntoList(env, null);
+    let dsl = new OpcodeBuilderDSL(buffer, null, env);
+
+    dsl.block({ templates: null }, (dsl, BEGIN, END) => {
```

Comment:
Also, I thought we were going to move away from pojo args. is this the only shape the dsl sees?